File: Frontend/src/app/core/services/stockManagment/productCharacteristicService/product-characteristic.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { ProductCharacteristic } from '../../../../shared/models/StockManagment/ProductCharacteristic.model';
import { ProductCharacteristicDTO } from '../../../../shared/models/dto/stockManagmentDTO/ProductCharacteristic.dto';

@Injectable({
  providedIn: 'root'
})
export class ProductCharacteristicService {

  private apiUrl = 'http://localhost:8080/api/productCharacteristic';
  private http = inject(HttpClient);

  getProductCharacteristic():Observable<ProductCharacteristic[]>{
    return this.http.get<ProductCharacteristic[]>(`${this.apiUrl}/ListProductCharacteristics`);
  }

  addProductCharacteristic(productCharacteristic : ProductCharacteristicDTO):Observable<ProductCharacteristic>{
    console.log("ProductCharacteristic service is called to add a new ProductCharacteristic");
    return this.http.post<ProductCharacteristic>(`${this.apiUrl}/addProductCharacteristic`,productCharacteristic);
  }

  findProductCharacteristicById(id:number):Observable<ProductCharacteristic>{
    return this.http.get<ProductCharacteristic>(`${this.apiUrl}/find/${id}`);
  }

  editProductCharacteristic(productCharacteristic:ProductCharacteristicDTO , productCharacteristicId:number):Observable<ProductCharacteristic>{
    return this.http.put<ProductCharacteristic>(`${this.apiUrl}/update/${productCharacteristicId}`,productCharacteristic);
  }

  deleteProductCharacteristic(productCharacteristicId:number):Observable<string>{
    return this.http.delete<string>(`${this.apiUrl}/delete/${productCharacteristicId}`);
  }
 
  addListProductCharacteristic(characteristicIds : Number[],producId:number):Observable<ProductCharacteristic[]>{
    return this.http.post<ProductCharacteristic[]>
    (`${this.apiUrl}/addProductCharacteristicList/${producId}`,characteristicIds);
  }

  getProductCharacteristicsByProductId(productId:number):Observable<ProductCharacteristic[]>{
    return this.http.get<ProductCharacteristic[]>(`${this.apiUrl}/allProductCharacteristicsByProductId/${productId}`);
  }
}

File: store/src/main/java/com/example/store/Service/stockManagment/interfaces/ProductCharacteristicService.java
package com.example.store.Service.stockManagment.interfaces;


import com.example.store.DTO.stockManagment.ProductCharacteristicDTO;
import com.example.store.Model.StockMangement.Characteristic;
import com.example.store.Model.StockMangement.ProductCharacteristic;

import java.util.List;

public interface ProductCharacteristicService {

    ProductCharacteristic saveProductCharacteristic(ProductCharacteristicDTO productCharacteristic);
    ProductCharacteristic findProductCharacteristicById(Long productCharacteristicId);
    List<ProductCharacteristic> fetchProductCharacteristicList();
    ProductCharacteristic updateProductCharacteristic(ProductCharacteristicDTO productCharacteristic,Long productCharacteristicId);

    void deleteProductCharacteristicByProductId(Long productId);
    void deleteProductCharacteristicById(Long productCharacteristicId);

    List<ProductCharacteristic> saveProductCharacteristicList(List<Long> characteristicList, Long productId);
    List<ProductCharacteristic> findProductCharacteristicByProductId(Long productId);

}
